### ISSUE TITLE
final_eval does not pick the right imgsz

### ISSUE THREAD

**Issue by AndreaCaliandro** [ultralytics/ultralytics#16023]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Looks like that during in the last phase of the training, the `final_eval()` step wants to use the default image size of 640, rather than the size passed to the `model.train`. 

I realized it when building a custom model with a few fully connected layers in the head. I train the model with the MNIST dataset using an image size of 32.
The train and the model work fine when going trough each of the epochs, 
but at the final validation I get the error message below.
As you can see, the problem is a multiplication of two matrices that do not match, 
which is most probably caused by the input image size larger than 32.
I think the problem is that the `model.warmup` (validator.py:157) does not get the right `imgsz`.

Since the training goes trough the epochs, it saves the `best.pt` and `last.pt` weights.
When loading these weights and predicting an MNIST image, everything works fine.

```
Traceback (most recent call last):
  File "/notebooks/shared/andrea/train_custom_model.py", line 7, in <module>
    results = model.train(data="mnist", epochs=3, imgsz=32)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py", line 673, in train
    self.trainer.train()
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py", line 199, in train
    self._do_train(world_size)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/trainer.py", line 459, in _do_train
    self.final_eval()
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/models/yolo/classify/train.py", line 146, in final_eval
    self.metrics = self.validator(model=f)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/validator.py", line 157, in __call__
    model.warmup(imgsz=(1 if pt else self.args.batch, 3, imgsz, imgsz))  # warmup
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/nn/autobackend.py", line 627, in warmup
    self.forward(im)  # warmup
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/nn/autobackend.py", line 453, in forward
    y = self.model(im, augment=augment, visualize=visualize, embed=embed)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py", line 89, in forward
    return self.predict(x, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py", line 107, in predict
    return self._predict_once(x, profile, visualize, embed)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py", line 128, in _predict_once
    x = m(x)  # run
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/linear.py", line 117, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x394384 and 400x120)
```

### Environment

Ultralytics YOLOv8.2.2 🚀 Python-3.10.12 torch-2.4.0+cu121 CUDA:0 (NVIDIA A100-SXM4-80GB, 81051MiB)
Setup complete ✅ (64 CPUs, 503.8 GB RAM, 373.5/862.4 GB disk)

### Minimal Reproducible Example

Training
```
from ultralytics import YOLO


model = YOLO("./custom_model.yaml", task='classify')
print(model.info(detailed=True))

results = model.train(data="mnist", epochs=10, imgsz=32)
```
\
Prediction
```
import cv2


mnist_image = "path_to/mnist/test/0/10.png"
img = cv2.imread(mnist_image)

model = YOLO("./custom_model.yaml", task='classify').load("./runs/classify/train/weights/best.pt")
results = model.predict(img)
results[0].probs
```

\
custom_model.yaml
```
# Simple image classifier as in https://pytorch.org/tutorials/beginner/onnx/export_simple_model_to_onnx_tutorial.html

nc: 10
backbone:
  # [from, repeats, module, args]
  - [-1, 1, nn.Conv2d, [3, 6, 5]]
  - [-1, 1, nn.MaxPool2d, [2]]  
  - [-1, 1, nn.Conv2d, [6, 16, 5]]
  - [-1, 1, nn.MaxPool2d, [2]]
  - [-1, 1, nn.Flatten, [1, -1]]

head:
  - [-1, 1, nn.Linear, [400, 120]]
  - [-1, 1, nn.Linear, [120, 84]]
  - [-1, 1, nn.Linear, [84, 10]]
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
Probably because `args` is not passed when creating the validator for classification:

https://github.com/ultralytics/ultralytics/blob/88102eb5085a668fff0171fdaa6297af65c2d010/ultralytics/models/yolo/classify/train.py#L110

**Comment by AndreaCaliandro** [comment]
Thank you @Y-T-G, I think you got the right solution.
I'll close this issue as soon as your PR #16025 get merged.

**Comment by glenn-jocher** [comment]
Great news 😃! Your original issue may now be resolved ✅ in PR #16025. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
